src/main/java/org/openstreetmap/atlas/geography/atlas/pbf/loading/OsmPbfReader.java

```diff
@@ -0,0 +1,420 @@
+    {
+        // TODO - add statistics around what was dropped, skipped and processed.
+        final RelationBean bean = new RelationBean();
+        for (final RelationMember member : relation.getMembers())
+        {
+            final long memberIdentifier = member.getMemberId();
+            final EntityType memberType = member.getMemberType();
+            final String role = member.getMemberRole();
+
+            if (memberType == EntityType.Node)
+            {
+                if (this.builder.peek().node(memberIdentifier) != null)
+                {
+                    bean.addItem(memberIdentifier, role, ItemType.POINT);
+                }
+                else
+                {
+                    logger.warn(MISSING_MEMBER_MESSAGE, relation.getId(), EntityType.Node,
+                            memberIdentifier);
+                }
+            }
+            else if (memberType == EntityType.Way)
+            {
+                if (this.builder.peek().line(memberIdentifier) != null)
+                {
+                    bean.addItem(memberIdentifier, role, ItemType.LINE);
+                }
+                else
+                {
+                    logger.warn(MISSING_MEMBER_MESSAGE, relation.getId(), EntityType.Way,
+                            memberIdentifier);
+                }
+            }
+            else if (memberType == EntityType.Relation)
+            {
+                if (this.builder.peek().relation(memberIdentifier) != null)
+                {
+                    bean.addItem(memberIdentifier, role, ItemType.RELATION);
+                }
+                else
+                {
+                    logger.warn(MISSING_MEMBER_MESSAGE, relation.getId(), EntityType.Relation,
+                            memberIdentifier);
+                }
+            }
+            else
+            {
+                // A Bound should never be a Relation member; if this is the case, log and continue
+                logger.error("Invalid Bound member {} found for Relation {}", member.getMemberId(),
+                        relation.getId());
+            }
+        }
+
+        return bean;
+    }
+
+    /**
+     * Constructs a {@link PolyLine} given an OSM PBF {@link Way}. The {@link Way} object doesn't
+     * contain the coordinates of its geometry, only the references to the {@link Node}s
+     * identifiers. We need to look up the {@link Location} of each {@link Node} and re-construct
+     * the {@link PolyLine} manually.
+     *
+     * @param way
+     *            The {@link Way} for which to construct the {@link PolyLine}
+     * @return the constructed {@link PolyLine}
+     */
+    private PolyLine constructWayPolyLine(final Way way)
+    {
+        final List<WayNode> wayNodes = way.getWayNodes();
+        final List<Location> wayLocations = new ArrayList<>();
+        wayNodes.forEach(node -> wayLocations.add(getNodeLocation(node.getNodeId())));
```

Comment:
In the rare case of a malformed pbf file, is it possible that `getNodeLocation` returns nothing here?

Reply:
It is possible - either we filtered it out or the PBF is malformed. If that's the case, we throw a `CoreException` on line 278 below. Let me add more context to the Exception message there.